### PR TITLE
Return failure reason in http header for undelete

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Return failure reason in http response header for undelete. Since undelete is not exposed outside of the linked datacenter and should be an admin operation, we should always return the failure reason in the header.

### CODE REVIEW

File: ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java [hunk 1]
```diff
@@ -257,6 +256,10 @@ public void handlePut(RestRequest restRequest, RestResponseChannel restResponseC
       } else if (requestPath.matchesOperation(Operations.UNDELETE) && frontendConfig.enableUndelete) {
         // If the undelete is not enabled, then treat it as unrecognized operation.
         undeleteHandler.handle(restRequest, restResponseChannel, (r, e) -> {
+          if (e instanceof RouterException) {
+            e = new RestServiceException(e.getMessage(),
+                RestServiceErrorCode.getRestServiceErrorCode(((RouterException) e).getErrorCode()), true);
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:261]: sorry, I realize that setting this option whenever a failure reason needs to be sent creates a lot of duplicate code. Maybe a better solution would be to have a request-level internal key that can be set to true when that request should be allowed to receive error info (for any exception type).

THe approach in this PR should work, though!

Comment (justinlin-linkedin) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java]: This is a good idea, updated the PR to use an internal key.

File: ambry-api/src/main/java/com/github/ambry/rest/RestUtils.java [hunk 2]
```diff
@@ -297,6 +296,11 @@
      * The key for the {@link RequestPath} that represents the parsed path of an incoming request.
      */
     public static final String REQUEST_PATH = KEY_PREFIX + "request-path";
+
+    /**
+     * To be set to {@code true} if failures reason should be attached to frontend responses.
+     */
+    public static final String SEND_FAILURE_REASON = KEY_PREFIX + "send-failure-reason";
```
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/rest/RestUtils.java:303]: thanks for making this change! I can change some of the code inside of AmbryLI to use this instead of the old approach.

File: ambry-test-utils/src/main/java/com/github/ambry/router/InMemoryRouter.java [hunk 3]
```diff
@@ -335,15 +335,13 @@ public void setLifeVersion(short lifeVersion) {
...
-      } else if (undeletedBlobs.contains(blobId)) {
-        exception = new RouterException("Blob already undeleted", RouterErrorCode.BlobUndeleted);
       } else if (!deletedBlobs.contains(blobId)) {
         exception = new RouterException("Blob not deleted", RouterErrorCode.BlobNotDeleted);
       }
       undeletedBlobs.add(blobId);
       deletedBlobs.remove(blobId);
       if (notificationSystem != null) {
-        notificationSystem.onBlobDeleted(blobId, serviceId, null, null);
+        notificationSystem.onBlobUndeleted(blobId, serviceId, null, null);
```
Comment (zzmao) [inline ambry-test-utils/src/main/java/com/github/ambry/router/InMemoryRouter.java:344]: why is this changed to `onBlobUndeleted `?

Comment (justinlin-linkedin) [reply ambry-test-utils/src/main/java/com/github/ambry/router/InMemoryRouter.java:344]: This was a mistake, this is actually undelete function, but I copied the function from delete and forgot to change onBlobDeleted to onBlobUndeleted. Now I am changing it.
